### PR TITLE
Add stack_arrays in C

### PR DESCRIPTION
**Pull request by saniQQ**
- Add handling of stack_array decorator in C for numpy funcs: array, full, ones, zeros, empty:
  - `_init_stack_array` : function is called from `array_fill` and `copy_NumpyArray_Data` to do the initialization,
  - `arrayFill` : handles the initialization when calling the functions : full, ones, zeros, empty,
  - `copy_NumpyArray_Data` : handles the initialization when calling the functions: array;
 
- Activate 1d stack_array epyccel tests;
- Add 2d stack_array epyccel tests for C and Fortran.

### CODE REVIEW

File: pyccel/codegen/printing/ccode.py [hunk 1]
```diff
@@ -1134,29 +1134,57 @@ def _print_Assign(self, expr):
...
+        shape = lhs.shape
+        shape = [self._print(i) for i in shape]
+        shape = ", ".join(a for a in shape)
+        shape_dtype = self.find_in_dtype_registry('int', 8)
+
+        shape_init = "({}[]){{{}}}".format(shape_dtype, shape)
+        strides_init = "({}[{}]){{0}}".format(shape_dtype, len(lhs.shape))
+        if isinstance(buffer_array, Variable):
+            buffer_array = "{}.{}".format(self._print(buffer_array), dtype)
+        cpy_data = '{0} = (t_ndarray){{.{1}={2},\n .shape={3},\n .strides={4},\n .nd={5},\n .type={6},\n .is_view={7}}};\n'.format(self._print(lhs), dtype, buffer_array, shape_init, strides_init, len(lhs.shape), dtype, 'false')
```
Comment (MenoIy) [inline pyccel/codegen/printing/ccode.py:1183]: can you break this into multiple lines?

File: pyccel/codegen/printing/ccode.py [hunk 2]
```diff
@@ -1134,29 +1134,57 @@ def _print_Assign(self, expr):
...
+                else:
+                    cpy_data = "memcpy({0}.{2}, {1}, {0}.buffer_size);".format(lhs, dummy_array_name, dtype)
                 return  '%s%s\n' % (dummy_array, cpy_data)
 
         if isinstance(rhs, (NumpyFull)):
             code_init = ''
+            if expr.lhs.is_stack_array:
+                declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+                lenght = [self._print(i) for i in expr.lhs.shape]
+                lenght = '*'.join(lenght)
```
Comment (MenoIy) [inline pyccel/codegen/printing/ccode.py:1156]: you can group this in join
```Python
 lenght = '*'.join(self._print(i) for i in expr.lhs.shape)
```

**Review by MenoIy — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/ccode.py [hunk 3]
```diff
@@ -1134,29 +1134,59 @@ def _print_Assign(self, expr):
...
         rhs = self._print(rhs)
         return '{} = {};'.format(lhs, rhs)
 
+    def _init_stack_array(self, expr, buffer_array, empty=False):
+        lhs = expr.lhs
+        rhs = expr.rhs
+        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
+        shape = lhs.shape
+        shape = [self._print(i) for i in shape]
+        shape = ", ".join(a for a in shape)
```
Comment (MenoIy) [inline pyccel/codegen/printing/ccode.py:1175]: you can group this too

**Review by MenoIy — CHANGES_REQUESTED** [review]

**Review by MenoIy — CHANGES_REQUESTED** [review]
The printer of Assign now is looking more complicated because a lot of numpy array functions are managed there,  I suggest splitting or creating new functions that manage numpy element.

File: pyccel/codegen/printing/ccode.py [hunk 4]
```diff
@@ -1120,42 +1118,76 @@ def _print_Assign(self, expr):
         if isinstance(expr.rhs, FunctionCall) and isinstance(expr.rhs.dtype, NativeTuple):
             self._temporary_args = [VariableAddress(a) for a in expr.lhs]
             return '{};'.format(self._print(expr.rhs))
+        if isinstance(expr.rhs, (NumpyArray)):
+            return '{}'.format(self.print_NumpyArray(expr))
+        if isinstance(expr.rhs, (NumpyFull)):
+            return '{}'.format(self.print_NumpyFull(expr))
```
Comment (MenoIy) [inline pyccel/codegen/printing/ccode.py:1124]: just return self.print_ directly.

**Review by MenoIy — CHANGES_REQUESTED** [review]

**Review by MenoIy — APPROVED** [review]

File: pyccel/codegen/printing/ccode.py [hunk 5]
```diff
@@ -1120,42 +1118,76 @@ def _print_Assign(self, expr):
...
+                cpy_data = self._init_stack_array(expr, dummy_array_name)
             else:
-                return ''
-            return '{}\n'.format(code_init)
+                cpy_data = "memcpy({0}.{2}, {1}, {0}.buffer_size);".format(self._print(lhs), dummy_array_name, dtype)
+            return  '%s%s\n' % (dummy_array, cpy_data)
 
-        rhs = self._print(rhs)
-        return '{} = {};'.format(lhs, rhs)
+    def print_NumpyFull(self, expr):
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1156]: See above, please rename this function

File: pyccel/codegen/printing/ccode.py [hunk 6]
```diff
@@ -1120,42 +1118,76 @@ def _print_Assign(self, expr):
...
+        rhs = expr.rhs
+        lhs = expr.lhs
+        code_init = ''
+        if lhs.is_stack_array:
+            declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+            lenght = '*'.join(self._print(i) for i in lhs.shape)
+            buffer_array = "({declare_dtype}[{lenght}]){{}}".format(declare_dtype = declare_dtype, lenght=lenght)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1163]: Typo `lenght`->`length`

File: pyccel/codegen/printing/ccode.py [hunk 7]
```diff
@@ -1120,42 +1118,76 @@ def _print_Assign(self, expr):
...
+
+    def _init_stack_array(self, expr, buffer_array, empty=False):
+        lhs = expr.lhs
+        rhs = expr.rhs
+        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
+        shape = ", ".join(self._print(i) for i in lhs.shape)
+        declare_dtype = self.find_in_dtype_registry('int', 8)
+
+        shape_init = "({declare_dtype}[]){{{shape}}}".format(declare_dtype=declare_dtype, shape=shape)
+        strides_init = "({declare_dtype}[{lenght}]){{0}}".format(declare_dtype=declare_dtype, lenght=len(lhs.shape))
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1181]: typo `lenght`->`length`

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/ccode.py [hunk 8]
```diff
@@ -1120,42 +1118,76 @@ def _print_Assign(self, expr):
...
+        rhs = expr.rhs
+        lhs = expr.lhs
+        if rhs.rank == 0:
+            raise NotImplementedError(lhs + "=" + rhs)
+        dummy_array_name, _ = create_incremented_string(self._parser.used_names, prefix = 'array_dummy')
+        declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
+        arg = rhs.arg
+        if rhs.rank > 1:
+            arg = functools.reduce(operator.concat, arg)
```
[the remark this answers is not in the document]
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:1139]: Its used to flatten the array

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:1139]: Could you add a comment for this please

File: pyccel/codegen/printing/ccode.py [hunk 9]
```diff
@@ -1120,42 +1118,76 @@ def _print_Assign(self, expr):
...
-            dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
+        rhs = self._print(expr.rhs)
+        return '{} = {};'.format(lhs, rhs)
 
-            arg = rhs.arg
-            if rhs.rank > 1:
-                arg = functools.reduce(operator.concat, arg)
-            if isinstance(arg, Variable):
-                arg = self._print(arg)
+    def print_NumpyArray(self, expr):
```
[the remark this answers is not in the document]
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:1129]: I need the `Assign` expression to be passed to the function if i add the `_` the printer will treat it as a rhs and i won't have acess to the lhs

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:1129]: Ah sorry! I missed the fact that the `_` is missing. In which case you are right, this is not a dangerous case, but I don't think it is very readable as it is easy to miss that and then be surprised that the argument is not the expected type. If you could find a clearer name that would be great. In any case it needs a doc string

File: pyccel/codegen/printing/ccode.py (hunk 5 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:1156]: This also needs a docstring

File: pyccel/codegen/printing/ccode.py [hunk 10]
```diff
@@ -1120,42 +1118,76 @@ def _print_Assign(self, expr):
...
+            code_init += self._init_stack_array(expr, buffer_array)
+        if rhs.fill_value is not None:
+            if isinstance(rhs.fill_value, Literal):
+                dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+                code_init += 'array_fill(({0}){1}, {2});\n'.format(dtype, self._print(rhs.fill_value), self._print(lhs))
+            else:
+                code_init += 'array_fill({0}, {1});\n'.format(self._print(rhs.fill_value), self._print(lhs))
+        return '{}'.format(code_init)
+
+    def _init_stack_array(self, expr, buffer_array, empty=False):
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1173]: Please add a docstring to this function. Codacy doesn't complain in this file as all the `_print_X` functions would have nearly the same docstring, but any other functions should be properly documented

File: pyccel/codegen/printing/ccode.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:1129]: Will try to make it more clear and add a doc string 👍

[the remark this answers is not in the document]
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py]: i changed the name to `copy_NumpyArray_Data`

File: pyccel/codegen/printing/ccode.py (hunk 5 above)
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py]: i changed the name to `arrayFill`

File: pyccel/codegen/printing/ccode.py [hunk 11]
```diff
@@ -1120,42 +1118,115 @@ def _print_Assign(self, expr):
         if isinstance(expr.rhs, FunctionCall) and isinstance(expr.rhs.dtype, NativeTuple):
             self._temporary_args = [VariableAddress(a) for a in expr.lhs]
             return '{};'.format(self._print(expr.rhs))
+        if isinstance(expr.rhs, (NumpyArray)):
+            return self.print_NumpyArray(expr)
+        if isinstance(expr.rhs, (NumpyFull)):
+            return self.print_NumpyFull(expr)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1124]: Don't the calls need updating too?

File: pyccel/codegen/printing/ccode.py [hunk 12]
```diff
@@ -1120,42 +1118,115 @@ def _print_Assign(self, expr):
...
-            if isinstance(arg, Variable):
-                arg = self._print(arg)
+        parameters
+        ----------
+            expr : PyccelAstNode
+                The Assign Node used to get the lhs and rhs
+        Return
+        ------
+            String
+                Return a str that contain the declaration of a dummy data_buffer and
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1139]: Grammar `contain`->`contains`

File: pyccel/codegen/printing/ccode.py (hunk 11 above)
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:1124]: my bad

File: pyccel/codegen/printing/ccode.py [hunk 13]
```diff
@@ -1120,42 +1118,115 @@ def _print_Assign(self, expr):
...
+                       a call to memcpy to copy it to NdArray struct
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1140]: Maybe you can be more general:
```
and a call to an operator which copies it to an NdArray struct
```
That way the docstring doesn't need updating when we fix issue #682

File: pyccel/codegen/printing/ccode.py [hunk 14]
```diff
@@ -1120,42 +1118,115 @@ def _print_Assign(self, expr):
...
+
+        Parameters
+        ----------
+            expr : PyccelAstNode
+                The Assign Node used to get the lhs and rhs
+            buffer_array : String
+                The data buffer
+        Returns
+        -------
+            Returns a string that contain the initialization of a stack_array
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1210]: Grammar `contain`->`contains`

File: pyccel/codegen/printing/ccode.py [hunk 15]
```diff
@@ -1120,42 +1118,115 @@ def _print_Assign(self, expr):
...
+        if rhs.fill_value is not None:
+            if isinstance(rhs.fill_value, Literal):
+                dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+                code_init += 'array_fill(({0}){1}, {2});\n'.format(dtype, self._print(rhs.fill_value), self._print(lhs))
+            else:
+                code_init += 'array_fill({0}, {1});\n'.format(self._print(rhs.fill_value), self._print(lhs))
+        return '{}'.format(code_init)
+
+    def _init_stack_array(self, expr, buffer_array):
+        """ return a string that have handle the assign of a stack ndarray
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1200]: Grammar `return a string that have handle the assign of a stack ndarray`->`return a string which handles the assignment of a stack ndarray`

File: pyccel/codegen/printing/ccode.py [hunk 16]
```diff
@@ -1120,42 +1118,115 @@ def _print_Assign(self, expr):
...
             else:
-                return ''
-            return '{}\n'.format(code_init)
+                cpy_data = "memcpy({0}.{2}, {1}, {0}.buffer_size);".format(self._print(lhs), dummy_array_name, dtype)
+            return  '%s%s\n' % (dummy_array, cpy_data)
 
-        rhs = self._print(rhs)
-        return '{} = {};'.format(lhs, rhs)
+    def arrayFill(self, expr):
+        """ print the Assignement of a NdArray
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1171]: Typo `Assignement`->`assignment`

File: pyccel/codegen/printing/ccode.py [hunk 17]
```diff
@@ -1120,42 +1118,115 @@ def _print_Assign(self, expr):
...
+
+        parameters
+        ----------
+            expr : PyccelAstNode
+                The Assign Node used to get the lhs and rhs
+        Return
+        ------
+            String
+                Return a str that contain a call to the C function array_fill,
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1180]: Grammar `contain`->``contains`

**Review by mostafaMamoni — APPROVED** [review]

**Review by EmilyBourne — APPROVED** [review]

File: pyccel/codegen/printing/ccode.py [hunk 18]
```diff
@@ -1146,45 +1144,117 @@ def _print_Assign(self, expr):
         if isinstance(expr.rhs, FunctionCall) and isinstance(expr.rhs.dtype, NativeTuple):
             self._temporary_args = [VariableAddress(a) for a in expr.lhs]
             return '{};'.format(self._print(expr.rhs))
+        if isinstance(expr.rhs, (NumpyArray)):
+            return self.copy_NumpyArray_Data(expr)
+        if isinstance(expr.rhs, (NumpyFull)):
+            return self.arrayFill(expr)
+        if isinstance(expr.rhs, NumpyArange):
+            return self.print_NumpyArange(expr.rhs, expr.lhs)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1152]: I really should have made the same comment on @MenoIy 's PR. :/

Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:1152]: > _EmilyBourne_
> Functions in pyccel always follow the rule that if their name is _print_X then the argument expr is of type X. Not following this rule is dangerous as this function may be called when you do not want it to be. For example:
> ```
> a = np.array([1,2,3])*4
> ```
> Please rename this function
>  
> _saniQQ_
> I need the Assign expression to be passed to the function if i add the _ the printer will treat it as a rhs and i won't have acess to the lhs
> 
> _EmilyBourne_
> Ah sorry! I missed the fact that the _ is missing. In which case you are right, this is not a dangerous case, but I don't think it is very readable as it is easy to miss that and then be surprised that the argument is not the expected type. If you could find a clearer name that would be great. In any case it needs a doc string


> Please add a docstring to this function. Codacy doesn't complain in this file as all the _print_X functions would have nearly the same docstring, but any other functions should be properly documented

Comment (MenoIy) [reply pyccel/codegen/printing/ccode.py:1152]: ah sorry, @saniQQ  can you please rename the function?
```Python
        """ print the assignment of a NumpyArange
        parameters
        ----------
            expr : NumpyArange
                The node holding NumpyArange
            lhs : Variable
                 The left hand of Assign
        Return
        ------
            String
                Return string that contains the Assign code and the For loop
                responsible for filling the array values
        """
```
feel free to update this doc
and for the name you can name it fill_NumpyArange or you can pick a better name

**Review by EmilyBourne — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**